crates/noirc_frontend/src/monomorphisation/mod.rs

```diff
@@ -0,0 +1,576 @@
+use std::collections::{BTreeMap, HashMap, VecDeque};
+
+use crate::{
+    hir_def::{
+        expr::*,
+        function::{FuncMeta, Parameters},
+        stmt::{HirAssignStatement, HirLValue, HirPattern, HirStatement},
+    },
+    node_interner::{self, NodeInterner, StmtId},
+    util::vecmap,
+    FunctionKind, IsConst, TypeBinding, TypeBindings,
+};
+
+use self::ast::{DefinitionId, FuncId, Functions};
+
+pub mod ast;
+pub mod printer;
+
+struct Monomorphiser {
+    // Store monomorphised globals and locals separately,
+    // only locals are cleared on each function call and only globals are monomorphised.
+    // Nested HashMaps in globals lets us avoid cloning HirTypes when calling .get()
+    globals: HashMap<node_interner::FuncId, HashMap<HirType, FuncId>>,
+    locals: HashMap<node_interner::DefinitionId, DefinitionId>,
+
+    /// Queue of functions to monomorphise next
+    queue: VecDeque<(node_interner::FuncId, FuncId, TypeBindings)>,
+
+    interner: NodeInterner,
+
+    next_local_id: u32,
+    next_function_id: u32,
+}
+
+type HirType = crate::Type;
+
+pub fn monomorphise(main: node_interner::FuncId, interner: NodeInterner) -> Functions {
+    let mut monomorphiser = Monomorphiser::new(interner);
+    let mut functions = monomorphiser.compile_main(main);
+
+    while !monomorphiser.queue.is_empty() {
+        let (next_fn_id, new_id, bindings) = monomorphiser.queue.pop_front().unwrap();
+        monomorphiser.locals.clear();
+
+        perform_instantiation_bindings(&bindings);
```

Comment:
The TypeVariables themselves are reference-counted objects so no structure is directly owning them. They're contained within some Types but that is about it.

Reference-counting them is fine since the occurs check in type checking prevents cycles from forming in types or type variables.